scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1749 @@
+	 * at least once within a targeted latency.
+	 */
+	share = calc_slice_share(p, taskc);
+	slice = (share * nr_cpus_onln) *
+		(LAVD_TARGETED_LATENCY_NS / cutil_cur->load_ideal);
+
+	/*
+	 * Take the task's greedy ratio into consideration. We assign a shorter
+	 * time slice when the task is greedy but not latency-critical.
+	 */
+	gr_ft = calc_greedy_factor(taskc);
+	slice = (slice * 1000) / gr_ft;
+
+	/*
+	 * Keep the slice in [LAVD_SLICE_MIN_NS, LAVD_SLICE_MAX_NS].
+	 */
+	if (slice < LAVD_SLICE_MIN_NS)
+		slice = LAVD_SLICE_MIN_NS;
+	else if (slice > LAVD_SLICE_MAX_NS)
+		slice = LAVD_SLICE_MAX_NS;
+
+	/*
+	 * If a task has yet to be scheduled (i.e., a freshly forked task or a
+	 * task just under sched_ext), don't give a fair amount of time slice
+	 * until knowing its properties. This helps to mitigate potential
+	 * system starvation caused by massively forking tasks (i.e., fork-bomb
+	 * attacks).
+	 */
+	if (!have_scheduled(taskc))
+		slice >>= 2;
+
+	taskc->slice_ns = slice;
+	return slice;
+}
+
+static bool transit_task_stat(struct task_ctx *taskc, int tgt_stat)
+{
+	/*
+	 * Update task loads only when the state transition is valid.
+	 *
+	 * initial state
+	 * -------------
+	 *     |
+	 *    \/
+	 * [STOPPING] --> [ENQ] --> [RUNNING]
+	 *    /\                        |
+	 *    |                         |
+	 *    +-------------------------+
+	 */
+	const static int valid_tgt_stat[] = {
+		LAVD_TASK_STAT_ENQ,
+		LAVD_TASK_STAT_RUNNING,
+		LAVD_TASK_STAT_STOPPING,
+	};
+	int src_stat = taskc->stat;
+
+	if (src_stat < _LAVD_TASK_STAT_MIN || src_stat > _LAVD_TASK_STAT_MAX) {
+		scx_bpf_error("Invalid task state: %d", src_stat);
+		return false;
+	}
+
+	if (valid_tgt_stat[src_stat] == tgt_stat) {
+		taskc->stat = tgt_stat;
+		return true;
+	}
+
+	return false;
+}
+
+static void update_stat_for_enq(struct task_struct *p, struct task_ctx *taskc,
+				struct cpu_ctx *cpuc)
+{
+	/*
+	 * Reflect task's load immediately.
+	 */
+	taskc->load_actual = calc_task_load_actual(taskc);
+	cpuc->load_actual += taskc->load_actual;
+	cpuc->load_ideal  += get_task_load_ideal(p);
+}
+
+static void update_stat_for_run(struct task_struct *p, struct task_ctx *taskc,
+				struct cpu_ctx *cpuc)
+{
+	u64 now, wait_period, interval;
+
+	/*
+	 * Since this is the start of a new schedule for @p, we update run
+	 * frequency in a second using an exponential weighted moving average.
+	 */
+	now = bpf_ktime_get_ns();
+	wait_period = now - taskc->last_stop_clk;
+	interval = taskc->run_time_ns + wait_period;
+	taskc->run_freq = calc_avg_freq(taskc->run_freq, interval);
+
+	/*
+	 * Update task state when starts running.
+	 */
+	taskc->last_start_clk = now;
+}
+
+static void update_stat_for_stop(struct task_struct *p, struct task_ctx *taskc,
+				 struct cpu_ctx *cpuc)
+{
+	u64 now, run_time_ns;
+
+	/*
+	 * If the task was runnable before governed by this scheduler, ignore
+	 * this run.
+	 */
+	now = bpf_ktime_get_ns();
+	if (now < taskc->last_start_clk)
```

Comment:
Here too, I don't think this can actually happen.

Reply:
You are right. I will remove the unnecessary condition check.